lib/typing/rinfer.ml

```diff
@@ -334,15 +342,340 @@ let rec infer hes env top =
           if !Typing.show_refinement then
             print_hes hes
-          else 
+          else
+            ()
+        | Error(s) -> Printf.printf "%s\n" s
+      end;
+      `Sat
+  | `Unsat -> `Unsat
+  | `Fail -> `Fail
+  | `Unknown -> `Unknown
+
+
+let print_refinements (constraints: (int * [> `Int ] Id.t list * refinement) list) =
+  print_endline "[Refinements]";
+  List.iter (fun (i, ids, ref) -> (
+    Printf.printf "[constraint %d]\n%s\n" i ([%derive.show: [> `Int ] Id.t list] ids);
+    (* List.iter (fun x -> Printf.printf "%s " (Id.show  x)) ids; *)
+    print_refinement ref;
+    print_newline ();
+    ()
+  )) constraints;
+  print_endline "[Refinements end]"
+
+
+let print_env (env: Rtype.t IdMap.t) =
+  print_string "[print_env:start]\n";
+  Rethfl_syntax.IdMap.iteri env ~f:(fun ~key:key ~data:data -> (
+    Printf.printf "%s[ID=%d]: " key.name key.id;
+    print_rtype data;
+    print_newline ()
+  ));
+  print_string "[print_env:end]\n"
+
+type annotation_config = {
+  annotated_func: string;
+  annotated_type: Rtype.t; [@opaque]
+  dependencies_annotated_func: string list;
+  dependencies_toplevel: string list;
+}
+[@@deriving show]
+
+
+let rec string_of_rty_skeleton (rty: Rtype.t) =
+    match rty with
+    | RArrow(x, y) -> "RArrow("^ (string_of_rty_skeleton x)^", "^( string_of_rty_skeleton y)^")"
+    | RBool(_) -> "RBool(RTrue)"
+    | RInt(_) -> "RInt(RId Rethfl_syntax.Id.(gen ~name:\"x\" `Int))"
+
+
+let dependent_funs f (hes : Rhflz.hes) =
+  let fvs f =
+    let { Rhflz.body; _ } = List.find (fun { Rhflz.var; _ } -> Id.eq f var) hes in
+    Rhflz.fvs body
+  in
+  let rec aux checked rest =
+    match IdSet.choose rest with
+    | None -> checked
+    | Some f ->
+      let fvs = fvs f in
+      let rest' = IdSet.remove (IdSet.diff fvs checked) f in
+      let checked' = IdSet.add checked f in
+      aux checked' rest'
+  in
+  aux IdSet.empty (fvs f)
+
+let annotation_of file (hes : hes) =
+  let top_funs = List.map (fun { var; _ } -> var) hes in
+  let open Rethfl_util in
+  let annotated_func,annotated_type =
+    In_channel.with_file file
+      ~f:(fun cin ->
+          cin
+          |> Lexing.from_channel
+          |> Parser.annot Lexer.token)
+  in
+  let annotated_func' = List.find_exn top_funs ~f:(fun f -> f.name = annotated_func) in
+  let dependencies_annotated_func =
+    dependent_funs annotated_func' hes
+    |> IdSet.elements
+    |> List.map ~f:(fun f -> f.Id.name)
+  in
+  let dependencies_toplevel =
+    top_funs
+    |> List.filter_map ~f:(fun var -> if Id.eq var annotated_func' then None else Some var.name)
+  in
+  {
+    annotated_func;
+    annotated_type;
+    dependencies_annotated_func;
+    dependencies_toplevel
+  }
+(*
+let annotation_of file hes =
+  let r = annotation_of file hes  in
+  Format.printf "annotation: %a@." pp_annotation_config r;
+  r
+*)
+
+(* infer with annotations *)
+let infer_based_on_annottations hes (env: Rtype.t IdMap.t) top file =
```

Comment:
There are so many duplicated codes that are also used in `infer` and `check_annotation`.
Probably we should clean this up